### PR TITLE
EventDispatcher hooks for patient create, and patient update

### PR DESCRIPTION
**Pull request by kchapple**
Questions:

- I didn't add this behavior to the "old style" new patient form. Should that be removed?
- (1) below may break API even though I think it's a bug. When returning the ProcessResult, the service would not append the pid, and uuid to the existing data attribute array of the Result. Instead, the [pid, uuid] array was placed in a NEW array as an element of the existing data attribute. 

In the PR:

1. Fixed issue in PatientService when passing array to addData()
2. Added escape_table to BaseService when getting uuid
3. Consolidated patient update and create to the PatientService
4. Added events so modules can subscribe to patient update and create events
5. Added example module for usage of patient update/create hooks

Fixes #

### CODE REVIEW

**Comment by kchapple** [comment]
"Question 2" from above is why API tests are failing.

**Comment by kchapple** [comment]
I definitely don't want to break compatibility and can change the code back. Just want to make sure that this is intentional because it seems like an API consumer would just want the key/value pairs and not an array of arrays of key/value pairs.

For example rather than this:
`'data' = [ 'pid' => 123, 'uuid' => 'some-long-string' ]`

We are returning something this:
`'data' = [ 0 => [ 'pid' => 123, 'uuid' => 'some-long-string' ] ]`

**Comment by kchapple** [comment]
Perhaps the correct resolution is to call setData() with an array, rather than addData() in the patient service's insert method, but I'd love to hear from one of the developers who knows the reasoning here. Thanks!

**Comment by sjpadgett** [comment]
> For example rather than this:
> `'data' = [ 'pid' => 123, 'uuid' => 'some-long-string' ]`
> 
> We are returning something this:
> `'data' = [ 0 => [ 'pid' => 123, 'uuid' => 'some-long-string' ] ]`

I need to look into this as I just had to touch this for errors however, api return doesn't look correct.

**Comment by kchapple** [comment]
@sjpadgett Check out the insert method on the PatientService where addData is called with an array. I'm not clear on the purpose of "mulit-line" returns, but that may be a feature of the API and not a bug. not sure.

**Comment by sjpadgett** [comment]
@kchapple 
The addData() is to send back the insert query results, good or bad.
The array returned is correct where validation results is sent in one array and result values the other.

**Comment by zerai** [comment]
> Second question from above is why API tests are failing.

The behavior in ProcessingResult.php (addData method) is changed
You should revert the changes I guess

File: library/patient.inc [hunk 1]
```diff
@@ -1044,61 +1044,38 @@ function pdValueOrNull($key, $value)
...
-        sqlStatement($sql);
+        $uuid = $patientService->getUuid($pid);
+        $uuidString = UuidRegistry::uuidToString($uuid);
+        $result = $patientService->update($uuidString, $new);
     }
 
-    return $db_id;
+    // From the returned ProcessingResult (contains data and error messages)
+    // retrieve the data and return the pid
+    $data = $result->getData();
```
Comment (sjpadgett) [inline library/patient.inc:1075]: I think you only need
$pid = $result->getData()[0]['pid'];

**Comment by kchapple** [comment]
@zerai @sjpadgett 

> 
> The addData() is to send back the insert query results, good or bad.
> The array returned is correct where validation results is sent in one array and result values the other.

Before my change the insert() method on the PatientService used addData() with an array passed to it like this:
`$processResult->addData([ 'pid' => 123, 'uuid' => 'some-long-string' ]);`
This resulted in a return value in the "data" part of the response like this:
` 'data' = [ 0 => [ 'pid' => 123, 'uuid' => 'some-long-string' ] ]`

So in my code, you're correct, I'd need to write:
`$pid = $result->getData()[0]['pid'];`

But will there every be more indexes that "0" or is this an error? If there will be, would there need to be some sort of search for the value I actually want to dig it out of the array of arrays? I guess my point is, I'm not sure why you would need a multi-dimensional array in the data portion of the response returned from the API. I guess I'm saying if it's intentional, that's fine, but if it is a bug, it should be fixed in the API and the test. I'm thinking it's a bug and it should be 
`$processResult->setData([ 'pid' => 123, 'uuid' => 'some-long-string' ]);`
not addData() ...and that the addData() method should take a key/value pairs to push on the array, and not create an entirely new array every time addData() is called. 

Thoughts?

Ken

**Comment by sjpadgett** [comment]
I see your point. 


> I'm thinking it's a bug and it should be
> `$processResult->setData([ 'pid' => 123, 'uuid' => 'some-long-string' ]);`

I'm thinking you're correct. This existing pattern would be more appropriate in building arrays one patient at a time in a get.
I say change in insert method. I can't see anywhere in code except restcontrollers we use.

**Comment by kchapple** [comment]
I guess for now, I'll drop it and do the following and leave the API alone:

`$pid = $result->getData()[0]['pid'];`

**Comment by sjpadgett** [comment]
Okay but I can't imagine our apis doing a multi insert on a single request. Maybe future. Brady may have thoughts.

**Comment by kchapple** [comment]
It's not just the inset() method on the PatientService, though... I don't think. The same thing happens on the update, or anywhere addData() is called. Because before I changed it, the addData() method in ProcessingResult called only array_push() on whatever is passed into the method. So, if you passed in an array into addData() (like in the insert), that whole array will be appended as a new element of the existing data array, not merged into the existing array.  

Not sure what to do. Maybe I should leave it as-is, and do your [0] workaround until this can be discussed?

Ken

**Comment by kchapple** [comment]
If we can get @bradymiller  to chime in, I'll be happy to document the behavior in the code, cause it was pretty strange to me and something that needs to be documented in the API and in the code.

Ken

**Comment by kchapple** [comment]
If there were multiple inserts, you would expect multiple ProcessingResults returned, each with their own validation messages.

**Comment by sjpadgett** [comment]
> that whole array will be appended as a new element of the existing data array, not merged into the existing array.

Yes, but that is the intent. I think service author(Matthew I think) envisioned that you'd add patients one at a time building a multi array of say, ten patients, to return to the interface. Expecting this, client would iterate through resulting array. Right?

**Comment by kchapple** [comment]
> Also, the result array structure should never change as by now, we're stuck with it.
Yeah, we'd have to version the API, which isn't happening now. So, I guess I'll use the [0] work-around and document the code. The goal was to consolidate patient-creation to one method rather than having 3-4, this API thing was just a side-track.

**Comment by kchapple** [comment]
My goal was to consolidate patient-creation to a single method, rather than have 3-4 laying around. This API business was a sidetrack. I will use the [0] workaround as you suggested and document the code.

Thanks @sjpadgett !

**Comment by kchapple** [comment]
I think our convo got out-of-sync. I think we're on the same page though. Nobody is versioning the API at this time, so the [0] is the way to go. I'd love to know the intent though.

**Comment by tmccormi** [comment]
Do you think that the classes in the services directory should be purely reserved for the API use-case, and the patient DB creation should be handled by a separate class?

I think we use other services classes elsewhere in the codebase outside the API, for facilities for example. Not sure if this is a good pattern though. 

Ken

**Comment by kchapple** [comment]
Sorry the last comment was me, I get Tony's emails in my inbox and when I reply from email github thinks it's tony.

**Comment by kchapple** [comment]
> you'd add patients one at a time building a multi array of say, ten patients, to return to the interface. Expecting this, client would iterate through resulting array. 

This would make sense, but you'd expect there to be an array of ProcessingResults (not just data part) with all the validation messages for each insert, etc. The reason I think it's a bug is because a new instance of ProcessingResult is instantiated each call of insert or update, so if you did multiple inserts, they would each have their own instance of result and return the data inside the [0] element of the array. 

I think what I might do is move the actual creation of PID and SQL insert into another function for use withing the internal code, and leave the processingResult stuff for the API.

Ken

**Comment by sjpadgett** [comment]
That's fine @kchapple do what you feel necessary. Mean time, I need to address this.

I never intended the standard services to be used in the way it turned out they are being used interacting with our Rest apis.
Should be FHIR/Standard RestController->RestService->StandardService with implementing appropriate BaseService. 
The RestService would do the necessary id/pid/uuid handling to called core services...

**Comment by kchapple** [comment]
I created new databaseInsert() and databaseUpdate() functions on the Patient Service to work around the test failure, and the ugly [0] thing.

**Comment by kchapple** [comment]
@sjpadgett Let me know if you think this will work?

File: library/patient.inc [hunk 2]
```diff
@@ -1044,61 +1044,36 @@ function pdValueOrNull($key, $value)
...
-
-        // add a uuid for the new patient
-        $sql .= ", `uuid` ='" . add_escape_custom((new UuidRegistry(['table_name' => 'patient_data']))->createUuid()) . "'";
-
-        $db_id = sqlInsert($sql);
+    // Create instance of patient service
+    $patientService = new \OpenEMR\Services\PatientService();
+    if ($create === true ||
+        $pid === null) {
+        $result = $patientService->databseInsert($new);
```
Comment (sjpadgett) [inline library/patient.inc:1066]: I think needs to be `if ($create === true || $pid === null) {`

Not sure why actions doesn't like.

**Comment by sjpadgett** [comment]
@kchapple Looks good to me. If testing well for you, let me know. Not sure about the styling error...

File: library/patient.inc (hunk 2 above)
Comment (bradymiller) [reply library/patient.inc:1066]: `databseInsert` is misspelled, so guessing breaking here

File: library/patient.inc [hunk 3]
```diff
@@ -1044,61 +1044,36 @@ function pdValueOrNull($key, $value)
...
-
-            $sql .= ", `$key` = " . pdValueOrNull($key, $value);
-        }
-
-        // add a uuid for the new patient
-        $sql .= ", `uuid` ='" . add_escape_custom((new UuidRegistry(['table_name' => 'patient_data']))->createUuid()) . "'";
-
-        $db_id = sqlInsert($sql);
+    // Create instance of patient service
+    $patientService = new \OpenEMR\Services\PatientService();
```
Comment (bradymiller) [inline library/patient.inc:1063]: minor thing, but shouldn't we place this in a `use` block at top of script?

Comment (kchapple) [reply library/patient.inc]: yes, thanks!

File: library/patient.inc (hunk 2 above)
Comment (kchapple) [reply library/patient.inc]: Typo was fixed in a previous commit. Not sure what the style issue is. Might be the position of the logic operator as Jerry suggested?

**Comment by kchapple** [comment]
This is ready

**Comment by kchapple** [comment]
If the styling messages are correct, they can be fixed by automation. Let me know if they can't, and I'll fix them. Thanks!

File: library/patient.inc [hunk 4]
```diff
@@ -1044,61 +1045,36 @@ function pdValueOrNull($key, $value)
...
-        }
-
-        // add a uuid for the new patient
-        $sql .= ", `uuid` ='" . add_escape_custom((new UuidRegistry(['table_name' => 'patient_data']))->createUuid()) . "'";
-
-        $db_id = sqlInsert($sql);
+    // Create instance of patient service
+    $patientService = new PatientService();
+    if ($create === true ||
+        $pid === null) {
```
Comment (sjpadgett) [inline library/patient.inc:1066]: @mdsupport 
sounds like actions wants this:
 if (
$create === true ||
        $pid === null
) {
or this:
 if ($create === true || $pid === null) {

actions won't auto fix. running the cs phar from command line will which is what many of us do before submitting PRs.
php phpcbf.phar -n --standard=ci/phpcs.xml library/patient.inc

a little too picky for my taste.

**Comment by stephenwaite** [comment]
be glad to bring it in and then bring in the autofix right after

**Comment by stephenwaite** [comment]
there are some easy to use dev tools in the https://github.com/openemr/openemr/blob/master/CONTRIBUTING.md

be nice to single out 1 script that's failing so don't need to run through the whole codebase, @bradymiller where is the devtools script located? in the docker?

**Comment by sjpadgett** [comment]
go for it.

**Comment by stephenwaite** [comment]
really nice @kchapple , thanks for another great contribution

**Comment by stephenwaite** [comment]
looks like found `devtoolsLibrary.source` in the root dir of the docker too

**Comment by bradymiller** [comment]
2DV

**Pull request merged**